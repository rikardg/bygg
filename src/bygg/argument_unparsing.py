import argparse
import functools
from typing import List


@functools.cache
def _build_dest_to_action(parser: argparse.ArgumentParser) -> dict:
    actions = parser._get_optional_actions() + parser._get_positional_actions()
    dest_to_action = {action.dest: action for action in actions}
    return dest_to_action


def _get_argument_for_dest(parser: argparse.ArgumentParser, dest: str) -> str | None:
    dest_to_action: dict[str, argparse.Action] = _build_dest_to_action(parser)
    action = dest_to_action.get(dest, None)
    if action is None:
        return None
    option_strings = action.option_strings
    return option_strings[-1] if option_strings else ""


def unparse_args(
    parser: argparse.ArgumentParser,
    args: argparse.Namespace,
    *,
    drop: List[str] | None = None,
) -> List[str]:
    """
    Convert an argparse.Namespace back to a list of command line arguments.

    Parameters
    ----------
    parser : argparse.ArgumentParser
        The parser that was used to parse the command line arguments.
    args : argparse.Namespace
        The parsed command line arguments to unparse.
    drop : List[str] | None, optional
        Which dest keys to drop, by default None.

    Returns
    -------
    List[str]
        A list of command line arguments
    """
    exec_list = []
    for k, v in vars(args).items():
        argument = _get_argument_for_dest(parser, k)
        if drop and k in drop:
            continue
        if v is False or v is None:
            # These are arguments that were not given
            continue
        if v is True:
            exec_list.append(argument)
        elif v:
            if argument:
                exec_list.append(
                    f"{argument}={','.join(v) if isinstance(v, (list, tuple)) else v}"
                )
            elif isinstance(v, (list, tuple)):
                exec_list.extend(v)
            else:
                assert False
        elif argument == "":
            continue
        else:
            # Could happen if we add another type of argument
            assert False
    return exec_list
