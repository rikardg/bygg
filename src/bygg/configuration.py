import os
import sys
from typing import Dict, List, Optional

import msgspec
import rich
import rich.status

PYTHON_INPUTFILE = "Byggfile.py"
YAML_INPUTFILE = "Byggfile.yml"


class SettingsSection(msgspec.Struct, forbid_unknown_fields=True):
    default_action: Optional[str] = None


class ActionItem(msgspec.Struct, forbid_unknown_fields=True):
    """
    name: The name of the action.
    description: A description of the action. Used in e.g. help messages.
    message: A message to print when the action is executed.
    inputs: A list of files that are used as input to the action.
    outputs: A list of files that are generated by the action.
    dependencies: A list of actions that must be executed before this action.
    is_entrypoint: Whether this action is an entrypoint. Entry points are actions that
    can be executed directly from the command line.
    shell: The shell command to execute when the action is run.
    """

    name: str
    description: Optional[str] = None
    message: Optional[str] = None
    inputs: Optional[List[str]] = None
    outputs: Optional[List[str]] = None
    dependencies: Optional[List[str]] = None
    is_entrypoint: Optional[bool] = None
    environment: Optional[str] = "default"
    shell: Optional[str] = None


class Environment(msgspec.Struct, forbid_unknown_fields=True):
    """
    name: The name of the environment.

    description: A description of the environment. Used in e.g. help messages.

    byggfiles: A list of Byggfiles that use this environment.

    inputs: A list of files that are used as input to the environment. Typically pip
    requirements files, but can be any files.

    venv_directory: The directory where the virtual environment is located. Will be
    recreated by Bygg if any of the inputs are modified.

    shell: The shell command for creating the environment.

    entrypoints: A list of entrypoints that use this environment.
    """

    byggfile: str
    inputs: List[str]
    venv_directory: str
    shell: str
    description: Optional[str] = None
    message: Optional[str] = None


class ByggFile(msgspec.Struct, forbid_unknown_fields=True):
    actions: List[ActionItem]
    settings: SettingsSection = msgspec.field(default_factory=SettingsSection)
    environments: Dict[str, Environment] = msgspec.field(default_factory=dict)


def load_python_build_file(build_file: str):
    if os.path.isfile(build_file):
        with open(build_file, "r") as f:
            # modify load path to make the current directory importable
            preamble = "import sys\nsys.path.insert(0, '.')\n\n"
            exec(preamble + f.read(), globals())


def read_config_file() -> ByggFile | None:
    if not os.path.isfile(YAML_INPUTFILE):
        return None

    try:
        with open(YAML_INPUTFILE, "r") as f:
            return msgspec.yaml.decode(f.read(), type=ByggFile)
    except Exception as e:
        rich.print(
            "[red bold]Error while reading configuration file "
            f"[yellow]{YAML_INPUTFILE}[/yellow]:[/red bold] {e}"
        )
        sys.exit(1)


def dump_schema():
    import json

    schema = msgspec.json.schema(ByggFile)
    print(json.dumps(schema, indent=2))
