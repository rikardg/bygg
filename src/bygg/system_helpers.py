"""
System-near helper functions. Only import things from the standard library here; this
way these helpers can be used also for utility scripts.
"""

import contextlib
import errno
import os
import pty
import shlex
import signal
import subprocess
from typing import Generator


class ProcessGenerator:
    """
    Class that wraps a process return value and a generator that yields the output of a
    process line by line.

    Attributes
    ----------
    returncode : int
        Process return code.
    generator : Generator
        Generator that yields the output of the process line by line.

    Yields
    ------
    str
        Output of the process.
    """

    returncode: int
    """Process return code."""

    generator: Generator
    """Generator that yields the output of the process line by line."""

    def __init__(self):
        self.returncode = -1

    def __iter__(self):
        self.value = yield from self.generator


def subprocess_tty(cmd, encoding="utf-8", timeout=10, **kwargs):
    """
    Wrapper around subprocess.Popen that sets up the process as if it were running in a
    TTY and returns a generator that yields the stdout of the process line by line.

    Parameters
    ----------
    cmd :
        Passed directly to subprocess.Popen; see its documentation for details.
    encoding : str, optional
        Encoding to use when reading stdout, by default "utf-8".
    timeout : int, optional
        Timeout to use when polling the child process to see if it has exited, by
        default 10 seconds.

    Returns
    -------
    ProcessGenerator
        Generator that yields the stdout of the child process line by line. After the
        generator is exhausted, the returncode attribute of the generator will be set to
        the exit code of the child process.

    Yields
    ------
    str
        Lines of stdout from the child process.

    Raises
    ------
    subprocess.TimeoutExpired
        Re-raised if the child process did not exit within the set timeout after having
        received both terminate and kill signals.

    Can also raise other exceptions from subprocess.Popen.
    """

    generator = ProcessGenerator()

    def subprocess_iterator():
        # From https://stackoverflow.com/a/77387332
        m, s = pty.openpty()
        p = subprocess.Popen(cmd, stdout=s, stderr=s, **kwargs)
        os.close(s)

        try:
            for line in open(m, encoding=encoding):
                if not line:  # EOF
                    break
                yield line
        except OSError as e:
            if e.errno != errno.EIO:  # EIO also means EOF
                raise
        finally:
            if p.poll() is None:
                p.send_signal(signal.SIGINT)
                try:
                    p.wait(timeout)
                except subprocess.TimeoutExpired:
                    p.terminate()
                    try:
                        p.wait(timeout)
                    except subprocess.TimeoutExpired:
                        p.kill()
                        raise
            p.wait()
            generator.returncode = p.returncode

    generator.generator = subprocess_iterator()
    return generator


def subprocess_tty_print(cmd, encoding="utf-8", timeout=10, **kwargs):
    """
    Thin wrapper around subprocess_tty that prints its output line by line. See
    subprocess_tty for details.


    Parameters
    ----------
    See subprocess_tty.

    Returns
    -------
    int
        The status code from the subprocess.
    """
    proc = subprocess_tty(cmd, encoding, timeout, **kwargs)
    for line in proc:
        print(line.rstrip())
    return proc.returncode


class ExitCode(int):
    """
    An int with a customised __bool__ method that considers 0 to be truthy. This is
    useful for working with subprocess return codes.
    """

    def __bool__(self):
        return self == 0


def call(cmd, encoding="utf-8", timeout=10, **kwargs) -> ExitCode:
    """
    Convenience wrapper around subprocess_tty_print; the command to be executed can be
    given as string that will be split instead of as an array. Output will be printed
    line by line.

    Parameters
    ----------
    cmd : str
        The command to be executed. Will be run through shlex.split().

    See subprocess_tty_print and subprocess_tty for the other parameters.

    Returns
    -------
    ExitCode
        The exit code from the subprocess. ExitCode is a subclass of int that can be
        used as a boolean; subprocess success (0) is considered truthy.

    Raises
    ------
    See subprocess_tty_print and subprocess_tty.
    """

    return ExitCode(subprocess_tty_print(shlex.split(cmd), encoding, timeout, **kwargs))


@contextlib.contextmanager
def change_dir(dir: str | os.PathLike | None):
    old_dir = os.getcwd()
    try:
        if dir is not None:
            os.chdir(dir)
        yield
    finally:
        os.chdir(old_dir)
