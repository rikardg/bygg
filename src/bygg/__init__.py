from .core.action import Action, ActionContext, WorkChannel, action
from .core.common_types import CommandStatus

__all__ = [
    "Action",
    "ActionContext",
    "CommandStatus",
    "WorkChannel",
    "action",
]
