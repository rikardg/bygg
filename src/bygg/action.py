from dataclasses import dataclass
from typing import Callable, Iterable, Literal, Optional, Set, Tuple

from bygg.common_types import CommandStatus

SchedulingType = Literal["in-process", "processpool"]

# Function that returns a string that is included in the dependency digest. Returning
# None causes the value to be ignored.
DynamicDependency = Callable[[], str | None]


@dataclass
class ActionContext:
    name: str
    message: str | None
    inputs: Set[str]
    outputs: Set[str]
    dependencies: Set[str]
    dynamic_dependency: Optional[DynamicDependency]
    is_entrypoint: bool
    scheduling_type: SchedulingType


Command = Callable[
    [ActionContext],
    CommandStatus,
]


class Action(ActionContext):
    """
    An action in the build graph.

    name: The name of the action.
    message (optional): A message to display when the action is run.
    inputs (optional): A list of paths that are inputs to the action.
    outputs (optional): A list of paths that are outputs of the action.
    dependencies (optional): A list of action names that this action depends on.
    is_entrypoint (optional): Whether this action is an entrypoint to the build graph.
    command (optional): Instance of one of the Command types to run.

    """

    command: Command | None
    dependency_files: Set[str]

    def __init__(
        self,
        name: str,
        description: str | None = None,
        message: str | None = None,
        inputs: Optional[Iterable[str]] = None,
        outputs: Optional[Iterable[str]] = None,
        dependencies: Optional[Iterable[str]] = None,
        dynamic_dependency: Optional[DynamicDependency] = None,
        is_entrypoint: bool = False,
        command: Command | None = None,
        scheduling_type: SchedulingType = "processpool",
    ):
        from bygg.scheduler import scheduler

        self.name = name
        self.description = description
        self.message = message
        self.inputs = {*inputs} if inputs else set()
        self.outputs = {*outputs} if outputs else set()
        self.dependencies = {*dependencies} if dependencies else set()
        self.dynamic_dependency = dynamic_dependency
        self.is_entrypoint = is_entrypoint
        self.command = command
        self.scheduling_type = scheduling_type

        self.dependency_files = set()

        scheduler.build_actions[name] = self

    def __repr__(self):
        return f"""Action(name={self.name}, inputs=({
                ', '.join([str(i) for i in self.inputs])}), outputs=({
                ', '.join([str(o) for o in self.outputs])}), dependencies=({
                ', '.join(self.dependencies)}) is_entrypoint={
            self.is_entrypoint})"""

    def __str__(self):
        return self.__repr__()


def action(
    name: str,
    *,
    message: Optional[str] = None,
    inputs: Optional[Iterable[str]] = None,
    outputs: Optional[Iterable[str]] = None,
    dependencies: Optional[Iterable[str]] = None,
    dynamic_dependency: Optional[DynamicDependency] = None,
    is_entrypoint: bool = False,
):
    """Decorator for creating an Action from a function."""

    def create_action(func: Callable):
        Action(
            name,
            message,
            inputs=inputs,
            outputs=outputs,
            dependencies=dependencies,
            dynamic_dependency=dynamic_dependency,
            is_entrypoint=is_entrypoint,
            command=func,
        )

    return create_action


def action_set(
    base_name: str,
    *,
    message: Optional[str] = None,
    file_pairs: Iterable[Tuple[str, str]],
    dependencies: Optional[Iterable[str]] = None,
    is_entrypoint: bool = False,
):
    """
    Decorator for creating individual Actions from a list of input-output file pairs and
    a function. Creates one Action that depends on all of the individual Actions. It has
    name = base_name, which in turn can be used as a dependency for other Actions.
    """

    def create_actions(func: Callable):
        action_list = []
        for input_file, output_file in file_pairs:
            action_name = f"{base_name}_{output_file}"
            action_list.append(action_name)
            Action(
                action_name,
                message,
                inputs=[input_file],
                outputs=[output_file],
                dependencies=dependencies,
                is_entrypoint=is_entrypoint,
                command=func,
            )
        Action(base_name, f"Action set {base_name}", dependencies=action_list)

    return create_actions
