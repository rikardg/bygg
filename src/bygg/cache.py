from dataclasses import dataclass
from pathlib import Path
import pickle

from bygg.scaffolding import STATUS_DIR, make_sure_status_dir_exists

DEFAULT_DB_FILE = STATUS_DIR / "cache.db"


@dataclass
class InputsOutputsDigests:
    inputs_digest: str
    outputs_digest: str
    dynamic_digest: str | None


@dataclass
class CacheState:
    digests: dict[str, InputsOutputsDigests]


class Cache:
    data: CacheState | None
    db_file: Path

    def __init__(self, db_file: Path = DEFAULT_DB_FILE):
        self.data = CacheState({})
        self.db_file = db_file

    def load(self):
        make_sure_status_dir_exists()
        try:
            with open(self.db_file, "rb") as f:
                self.data = pickle.load(f)
        except (EOFError, FileNotFoundError):
            self.data = CacheState({})

    def save(self):
        if not self.data:
            return
        with open(self.db_file, "wb") as f:
            pickle.dump(self.data, f)
        # print(f"Cache: {self.data.digests}")

    def get_digests(self, name: str) -> InputsOutputsDigests | None:
        if not self.data:
            return None
        return self.data.digests.get(name, None)

    def set_digests(
        self,
        name: str,
        inputs_digest: str,
        outputs_digest: str,
        dynamic_digest: str | None = None,
    ):
        if not self.data:
            return
        self.data.digests[name] = InputsOutputsDigests(
            inputs_digest, outputs_digest, dynamic_digest
        )

    def remove_digests(self, name: str):
        if not self.data:
            return
        self.data.digests.pop(name, None)
