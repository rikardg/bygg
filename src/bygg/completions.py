import argparse
from pathlib import Path
import shutil
import sys
import textwrap
from typing import Any, Set

from argcomplete.completers import DirectoriesCompleter
from argcomplete.finders import CompletionFinder

from bygg.output import output_plain


class ByggfileDirectoriesCompleter(DirectoriesCompleter):
    """
    A completer for directories that contain Bygg files.
    """

    def __call__(self, prefix, **kwargs):
        directories = super().__call__(prefix, **kwargs)
        byggfile_dirs = set()
        for dir in directories:
            byggfile_dirs |= {
                str(b.parent)
                for b in Path(dir).rglob("Byggfile.*")
                if b.suffix in {".py", ".yml"}
            }
        return sorted(list(byggfile_dirs))


def construct_already_there(
    parser: argparse.ArgumentParser | Any, options: Set[str]
) -> Set[str]:
    """
    Construct a set of options that are already present among the command line
    arguments. Adds both the short and long versions if they exist.
    """
    already_there_set: Set[str] = set()
    for action in parser._get_optional_actions():
        option_strings = set(action.option_strings)
        if option_strings & options:
            already_there_set.update(option_strings)
    return already_there_set


singular_options = {
    "--completions",
    "--dump-schema",
    "--help",
    "--version",
    "-h",
    "-v",
}

directory_option = {"-C", "--directory"}


class ByggCompletionFinder(CompletionFinder):
    """
    Override _get_completions because we want custom completion patterns that are not
    expressed in argparse.
    """

    def _get_completions(
        self, comp_words, cword_prefix, cword_prequote, last_wordbreak_pos
    ):
        comp_words_set = set(comp_words)

        # Stop completions if we already have a "singular" option, i.e. an option that
        # should not be followed by anything else.
        if comp_words_set & singular_options:
            return []

        # Get completions from argparse.
        completions = set(
            super()._get_completions(
                comp_words, cword_prefix, cword_prequote, last_wordbreak_pos
            )
        )

        # If the last word is a directory, the default completions will be from the
        # directories completer, so return these.
        if comp_words[-1] in directory_option:
            return completions

        action_completions = list(filter(lambda x: not x.startswith("-"), completions))
        already_there = construct_already_there(self._parser, comp_words_set)

        if comp_words_set & {"--tree", "--clean"}:
            # Actions if we have them.
            if action_completions and not cword_prefix.startswith("-"):
                return action_completions
            return completions & directory_option - already_there

        if comp_words_set & {"--list"}:
            # Only complete directories.
            return completions & directory_option - already_there

        # Return action completions if they exist and user has not entered a '-'.
        if action_completions and not cword_prefix.startswith("-"):
            return action_completions

        return completions - already_there


def do_completion(parser: argparse.ArgumentParser):
    completer = ByggCompletionFinder()
    completer(parser)


def generate_shell_completions(print_and_exit: bool):
    bygg_bin = Path(sys.argv[0])
    bygg_completions = bygg_bin.resolve().parent / "_bygg_completions.sh"
    if not bygg_completions.exists():
        import argcomplete.shell_integration

        with open(bygg_completions, "w") as f:
            f.write(argcomplete.shell_integration.shellcode([str(bygg_bin.name)]))

    if print_and_exit:
        terminal_cols, _ = shutil.get_terminal_size()
        indent_string = "# "
        width = min(terminal_cols, 80) - len(indent_string)
        description = f"""\
            A shell completions script has been generated in {bygg_completions}. It will
            pick up the bygg that is in PATH. Add the following line to your shell's
            startup script to load completions:
        """
        output_plain(
            textwrap.indent(
                textwrap.fill(
                    textwrap.dedent(description),
                    width=width,
                    break_long_words=False,
                    break_on_hyphens=False,
                ),
                indent_string,
            )
        )
        output_plain(f"\nsource {bygg_completions}")

        sys.exit(0)
