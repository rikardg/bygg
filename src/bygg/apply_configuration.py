from pathlib import Path
import shutil
import subprocess
import sys
import time
from typing import List

import rich
import rich.status

from bygg.action import Action
from bygg.configuration import (
    PYTHON_INPUTFILE,
    ByggFile,
    Environment,
    load_python_build_file,
)
from bygg.digest import calculate_string_digest
from bygg.output import output_error, output_info, output_plain
from bygg.scheduler import scheduler
from bygg.util import create_shell_command


def calculate_environment_hash(environment: Environment) -> str:
    requirements: List[str] = []
    requirements.append(environment.shell)
    if environment.inputs:
        for input in environment.inputs:
            with open(input, "r") as f:
                requirements += f.readlines()
    return calculate_string_digest(" ".join(requirements))


loading_python_build_file = rich.status.Status(
    "[cyan]Executing Python build file", spinner="dots"
)


def setup_environment(environment: Environment):
    venv_path = Path(
        environment.venv_directory if environment.venv_directory else ".venv"
    )

    environment_hash = calculate_environment_hash(environment)
    environment_hash_file = venv_path / "bygg_environment_hash.txt"
    if environment_hash_file.exists():
        with open(environment_hash_file, "r") as f:
            if f.read() == environment_hash:
                return True
        environment_hash_file.unlink()

    # TODO should we remove the whole venv or just trust pip to do the right thing? For
    # now, remove the venv.

    if venv_path.exists():
        output_info(f"Replacing venv at {venv_path}")
        shutil.rmtree(venv_path)

    output_info("Setting up environment")
    process = subprocess.run(
        environment.shell,
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        encoding="utf-8",
    )

    output_plain(process.stdout)

    if process.returncode != 0:
        output_error("Error while creating virtual environment:")
        output_plain(process.stdout)
        sys.exit(1)

    with open(environment_hash_file, "w") as f:
        f.write(environment_hash)

    return True


def should_restart_with(environment: Environment) -> str | None:
    venv_bin_path = Path(environment.venv_directory) / "bin"

    # # TODO: we could check here if the new venv is using the same Python binary that
    # # we're already running under, and if so just activate the venv.

    is_relative = Path(sys.executable).is_relative_to(venv_bin_path.resolve())
    if not is_relative:
        return str((venv_bin_path / "bygg"))
    return None


def apply_configuration(
    configuration: ByggFile | None,
    environment_name: str | None,
    is_restarted_with_env: str | None,
) -> str | None:
    """Returns the path of the bygg install to restart with if a restart is needed."""
    if configuration:
        if not is_restarted_with_env and environment_name:
            if environment_name in configuration.environments:
                environment = configuration.environments[environment_name]
                setup_environment(environment)

                restart_with = should_restart_with(environment)
                if restart_with is not None:
                    return restart_with

        for action in configuration.actions:
            # TODO Hack until we can iterate over the dependency graph in scheduler.prepare_run:
            if is_restarted_with_env and action.environment != is_restarted_with_env:
                continue

            shell_command = (
                create_shell_command(action.shell, action.message)
                if action.shell
                else None
            )
            Action(
                action.name,
                is_entrypoint=bool(action.is_entrypoint),
                inputs=action.inputs,
                outputs=action.outputs,
                dependencies=action.dependencies,
                command=shell_command,
            )

    # Evaluate the Python build file:

    python_build_file = PYTHON_INPUTFILE
    if configuration and environment_name:
        environment = configuration.environments.get(environment_name, None)
        if environment:
            python_build_file = environment.byggfile

    t0 = time.time()
    action_count = len(scheduler.build_actions)

    with loading_python_build_file:
        load_python_build_file(python_build_file)
    output_info(
        f"{len(scheduler.build_actions) - action_count} actions registered in "
        f"{time.time() - t0:.2f} seconds."
    )

    return None
