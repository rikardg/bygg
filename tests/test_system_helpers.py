import os

from bygg.system_helpers import (
    ExitCode,
    call,
    change_dir,
    subprocess_tty,
    subprocess_tty_print,
)
import pytest


def test_ExitCode():
    # Behaviour as a bool (inverted):
    assert ExitCode(0)
    assert not ExitCode(1)
    assert bool(ExitCode(0)) is True
    assert bool(ExitCode(1)) is False
    # Behaviour as an int:
    assert ExitCode(0) == 0
    assert ExitCode(0) != 1
    assert ExitCode(1) == 1
    assert ExitCode(1) != 0
    assert ExitCode(0) + 3 == 3
    assert ExitCode(1) + 3 == 4
    assert 3 + ExitCode(0) == 3
    assert 3 + ExitCode(1) == 4
    assert ExitCode(0) + ExitCode(0) == 0
    assert ExitCode(3) + ExitCode(2) == 5
    assert ExitCode(0) - ExitCode(0) == 0
    assert ExitCode(3) - ExitCode(2) == 1
    assert ExitCode(0) * ExitCode(0) == 0
    assert ExitCode(3) * ExitCode(2) == 6
    assert ExitCode(3) / ExitCode(2) == 1.5
    assert ExitCode(8) / ExitCode(-4) == -2

    with pytest.raises(ZeroDivisionError):
        assert ExitCode(0) / ExitCode(0) == 0


shell_command = "echo 'foo\nbar\nbaz'"


def test_call(capsys, snapshot):
    call(shell_command)
    captured = capsys.readouterr()
    assert captured.out == snapshot


def test_subprocess_tty_print(capsys, snapshot):
    subprocess_tty_print(shell_command.split(" "))
    captured = capsys.readouterr()
    assert captured.out == snapshot


def test_subprocess_tty():
    words = shell_command.split(" ")[-1].split("\n")
    for line, word in zip(subprocess_tty(shell_command.split(" ")), words):
        assert line.rstrip() == word


def test_change_dir():
    start_dir = os.getcwd()
    with change_dir("/"):
        assert os.getcwd() == "/"
    assert os.getcwd() == start_dir

    with change_dir("examples"):
        assert os.getcwd() == os.path.join(start_dir, "examples")
    assert os.getcwd() == start_dir

    with change_dir("examples/checks"):
        assert os.getcwd() == os.path.join(start_dir, "examples/checks")
    assert os.getcwd() == start_dir

    with change_dir("examples/trivial"):
        assert os.getcwd() == os.path.join(start_dir, "examples/trivial")
    assert os.getcwd() == start_dir

    with change_dir(None):
        assert os.getcwd() == start_dir
    assert os.getcwd() == start_dir
