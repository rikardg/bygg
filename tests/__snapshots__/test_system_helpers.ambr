# serializer version: 1
# name: test_call
  '''
  foo
  bar
  baz
  
  '''
# ---
# name: test_subprocess_tty_print
  '''
  'foo
  bar
  baz'
  
  '''
# ---
